### ISSUE TITLE
raft: clean unstable log early

### ISSUE THREAD

**Issue by pav-kv** [cockroachdb/cockroach#122438]
### Background

The [unstable](https://github.com/cockroachdb/cockroach/blob/06c9608ec395d130c7f86d95c5b763d770b13a15/pkg/raft/log_unstable.go#L36) log structure in `pkg/raft` holds log entries until they have been written to storage **and** fsync-ed.

After the [introduction](https://github.com/etcd-io/raft/pull/8) of async log writes, the flow of entries from memory to `Storage` is:

1. Entries are [appended](https://github.com/cockroachdb/cockroach/blob/06c9608ec395d130c7f86d95c5b763d770b13a15/pkg/raft/log.go#L143) to `unstable`.
2. On `handleRaftReady`, the `unstable` entries are [extracted](https://github.com/cockroachdb/cockroach/blob/06c9608ec395d130c7f86d95c5b763d770b13a15/pkg/kv/kvserver/replica_raft.go#L830) and [paired](https://github.com/cockroachdb/cockroach/blob/06c9608ec395d130c7f86d95c5b763d770b13a15/pkg/raft/rawnode.go#L169-L170) with a `MsgStorageAppend` message.
3. The batch of entries is [written](https://github.com/cockroachdb/cockroach/blob/06c9608ec395d130c7f86d95c5b763d770b13a15/pkg/kv/kvserver/replica_raft.go#L1065) to Pebble.
   - If async log writes are enabled, and the batch qualifies for an async write, the batch is written to Pebble, but [not synced](https://github.com/cockroachdb/cockroach/blob/06c9608ec395d130c7f86d95c5b763d770b13a15/pkg/kv/kvserver/logstore/logstore.go#L258-L261).
   - If the write doesn't qualify for async write, the entries are written **and** synced.
4. When the entries have been synced, we/Pebble invoke a callback which [sends](https://github.com/cockroachdb/cockroach/blob/06c9608ec395d130c7f86d95c5b763d770b13a15/pkg/kv/kvserver/replica_raft.go#L1613) a `MsgStorageAppendResp` responses back to the raft instance.
5. When [handling](https://github.com/cockroachdb/cockroach/blob/06c9608ec395d130c7f86d95c5b763d770b13a15/pkg/raft/raft.go#L1119) the append response, raft [removes](https://github.com/cockroachdb/cockroach/blob/06c9608ec395d130c7f86d95c5b763d770b13a15/pkg/raft/log_unstable.go#L158-L162) entries from `unstable` [^1].

### Improvement

In this flow, there is a period of time (between steps 3-4) when an entry has already been written to Pebble and sits in memtables, but still resides in the `unstable` struct. When async writes are enabled, this can last for multiple `Ready` iterations. Holding these entries in `unstable` is not strictly necessary, because they are already readable from the log `Storage`. We should clear them in step (3). This will, effectively, become a "transfer" of entries from `unstable` to `Storage`.

In Replication AC, entry tokens are [admitted](https://github.com/cockroachdb/cockroach/blob/06c9608ec395d130c7f86d95c5b763d770b13a15/pkg/kv/kvserver/replica_raft.go#L1059-L1061) and returned to the leader in step (3), too. Clearing the `unstable` entries at this point effectively includes them into the replication token "lifetime", and protects the node from OOMs caused by `unstable` build-ups.

The modification will be along the lines of having a new method/message to raft saying that some/all entries in `unstable` have been (non-durably) written, so raft can clear them. There can be some complications in the interaction with the async writes [protocol](https://github.com/cockroachdb/cockroach/blob/06c9608ec395d130c7f86d95c5b763d770b13a15/pkg/raft/rawnode.go#L277-L356).

Epic: CRDB-25503

[^1]: Some entries may have already been cleared from `unstable` by this time, e.g. if the leadership changed and the new leader has overwritten some entries. We only remove the entries that are a guaranteed to be matched by storage, and there are no in-flight appends overwriting them. See [this](https://github.com/cockroachdb/cockroach/blob/06c9608ec395d130c7f86d95c5b763d770b13a15/pkg/raft/rawnode.go#L277-L356) comment for some details.

**Comment by pav-kv** [comment]
@sumeerbhola With RACv2, this issue is still applicable, but to a lesser degree.

We admit / return tokens when the entry is both synced and admitted. So the "admission" happens at step 4 of [the flow](https://github.com/cockroachdb/cockroach/issues/122438#issue-2245899793), or later. After this, token returns to the leader are racing with step 5. If step 5 is delayed (e.g. the node is overloaded or has scheduling latency issue, or `Ready` handling is slow), the token return can outpace step 5, and a subsequent `MsgApp` is received before the `RawNode` was able to clear the synced entries from `unstable`. We now pile up more entries to `unstable`.

This is probably a rare case.

**Comment by sumeerbhola** [comment]
> @sumeerbhola With RACv2, this issue is still applicable, but to a lesser degree.

Agreed. I don't think we should bother doing something here unless we actually see a problem. AC (modulo deficiencies) is supposed to keep goroutine scheduling latency within some acceptable bound. And if the raft scheduler is clogged up since `Ready` handling is slow, say due to slow storage reads, this will also affect other ranges and so we would need to find a solution for that anyway.
